docs/spec_draft.md

```diff
@@ -703,6 +704,82 @@ implementation-defined.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.fft
+
+### Semantics
+
+Performs the forward and inverse Fourier Transforms for real and complex
+inputs/outputs. Multidimensional FFTs up to rank 3 are supported.
+
+`fft_type` is one of the following:
+
+* `FFT`: Forward complex-to-complex FFT. Shape is unchanged.
+* `IFFT`: Inverse complex-to-complex FFT. Shape is unchanged.
+* `RFFT`: Forward real-to-complex FFT. Shape of the innermost axis is reduced to
+`fft_length[-1] // 2 + 1` if `fft_length[-1]` is a non-zero value, omitting the
+reversed conjugate part of the transformed signal beyond the Nyquist frequency.
+* `IRFFT`: Inverse real-to-complex FFT (i.e. takes complex, returns real). Shape
+of the innermost axis is expanded to `fft_length[-1]` if `fft_length[-1]` is a
+non-zero value, inferring the part of the transformed signal beyond the Nyquist
+frequency from the reverse conjugate of the 1 to `fft_length[-1] // 2 + 1`
+entries.
+
+`fft_length` is the time-domain lengths of the axes being transformed. This is
+needed in particular for IRFFT to right-size the innermost axis, since `RFFT` of
+with even `fft_length` has the same output shape as odd `fft_length`.
+
+For multidimensional FFT, when more than one `fft_length` is provided, this is
+equivalent to applying a cascade of FFT operations to each of the innermost
+axes. Note that for the real->complex and complex->real cases, the innermost
+axis transform is (effectively) performed first (`RFFT`; last for `IRFFT` to
+have compatible element type), which is why the innermost axis is the one which
+changes size. Other axis transforms will then be complex->complex.
```

Comment:
As usual, the last two paragraphs need to be formalized to provide actionable instructions to implementors, and this is the first time when the formalization is significantly bigger than the prose that explains it:

```
More formally, given the function `fft` which takes 1-dimensional tensors of 
complex types as input, produces 1-dimensional tensors of same types as 
output and computes the discrete Fourier transform:

For `fft_type = FFT`, `result` is defined as the final result of a series of L 
computations where `L = size(fft_length)`. For example, for `L = 3`:
  * `result1[i0, ..., :]` = `fft(operand[i0, ..., :])` for all `i`.
  * `result2[i0, ..., :, iR-1]` = `fft(result1[i0, ..., :, iR-1])` for all `i`.
  * `result[i0, ..., :, iR-2, iR-1]` = `fft(result2[i0, ..., :, iR-2, iR-1])` 
    for all `i`.

Furthermore, given the function `ifft` which has the same type signature and 
computes the inverse of `fft`:

For `fft_type = IFFT`, `result` is defined as the inverse of the computations
for `fft_type = FFT`. For example, for `L = 3`:
  * `result1[i0, ..., :, iR-2, iR-1]` = `ifft(operand[i0, ..., :, iR-2, iR-1])` 
    for all `i`.
  * `result2[i0, ..., :, iR-1]` = `ifft(result1[i0, ..., :, iR-1])` for all `i`.
  * `result[i0, ..., :]` = `ifft(result2[i0, ..., :])` for all `i`.

Furthermore, given the function `rfft` which takes 1-dimensional tensors of
floating-point types, produces 1-dimensional tensors of complex types of the 
same floating-point semantics and works as follows:
  * `rfft(real_operand) = truncated_result` where
  * `complex_operand[i] = (real_operand, 0)` for all `i`.
  * `complex_result = fft(complex_operand)`.
  * `truncated_result = complex_result[:(rank(complex_result) / 2 + 1)]`.

(When the discrete Fourier transform is computed for real operands, the first
`N/2 + 1` elements of the result unambiguously define the rest of the result,
so the result of `rfft` is truncated to avoid computing redundant elements).

For `fft_type = RFFT`, `result` is defined as the final result of a series of L 
computations where `L = size(fft_length)`. For example, for `L = 3`:
  * `result1[i0, ..., :]` = `rfft(operand[i0, ..., :])` for all `i`.
  * `result2[i0, ..., :, iR-1]` = `fft(result1[i0, ..., :, iR-1])` for all `i`.
  * `result[i0, ..., :, iR-2, iR-1]` = `fft(result2[i0, ..., :, iR-2, iR-1])` 
    for all `i`.

Finally, given the function `irfft` which has the same type signature and 
computes the inverse of `rfft`:

For `fft_type = IRFFT`, `result` is defined as the inverse of the computations
for `fft_type = RFFT`. For example, for `L = 3`:
  * `result1[i0, ..., :, iR-2, iR-1]` = `ifft(operand[i0, ..., :, iR-2, iR-1])` 
    for all `i`.
  * `result2[i0, ..., :, iR-1]` = `ifft(result1[i0, ..., :, iR-1])` for all `i`.
  * `result[i0, ..., :]` = `irfft(result2[i0, ..., :])` for all `i`.
```

Reply:
Thank you for this, the explanation is lengthy yet concise and understandable.